[perfdhcp] use random macs from a list stored in a file.

pallotron opened the pull request:
I needed a way to perform stress tests using well defined mac addresses (this is to test our hook library that talks to an internal backend).

I added a -M <file_path> option, which is supposed to contain a list of mac addresses in text form with : to separate each couple of bytes.

dhcpv4 support is working and I've tested it.

Each requests uses a random mac selected via the `isc::util::random::UniformRandomIntegerGenerator` class to have some uniformity. 

dhcpv6 support is in the making, I have made an assumption to use DUID_LL when generating DUIDs (using random mac addresses in the list).

pallotron commented:
```
$ wc -l /home/pallotron/test.macs.txt
1000 /home/pallotron/test.macs.txt
[Exit code 0 @ 04:56:48]

$  sudo ./src/bin/perfdhcp/perfdhcp -4 -r 200 -p 40 -t 1 -i -M ~/test.macs.txt 10.92.107.21
Running: perfdhcp -4 -r 200 -p 40 -t 1 -i -M /home/pallotron/test.macs.txt 10.92.107.21
sent: 196; received: 159; drops: 37
sent: 393; received: 368; drops: 25
sent: 590; received: 563; drops: 27
sent: 787; received: 779; drops: 8
sent: 984; received: 980; drops: 4
sent: 1181; received: 1172; drops: 9
sent: 1378; received: 1378; drops: 0
sent: 1575; received: 1571; drops: 4
sent: 1772; received: 1756; drops: 16
sent: 1969; received: 1962; drops: 7
sent: 2166; received: 2166; drops: 0
sent: 2363; received: 2361; drops: 2
sent: 2560; received: 2555; drops: 5
sent: 2757; received: 2757; drops: 0
sent: 2955; received: 2950; drops: 5
sent: 3152; received: 3152; drops: 0
sent: 3349; received: 3349; drops: 0
sent: 3546; received: 3546; drops: 0
sent: 3743; received: 3740; drops: 3
sent: 3940; received: 3940; drops: 0
sent: 4137; received: 4137; drops: 0
sent: 4334; received: 4333; drops: 1
sent: 4531; received: 4529; drops: 2
sent: 4728; received: 4728; drops: 0
sent: 4924; received: 4920; drops: 4
sent: 5121; received: 5121; drops: 0
sent: 5318; received: 5318; drops: 0
sent: 5515; received: 5515; drops: 0
sent: 5712; received: 5712; drops: 0
sent: 5908; received: 5908; drops: 0
sent: 6105; received: 6105; drops: 0
sent: 6302; received: 6302; drops: 0
sent: 6499; received: 6497; drops: 2
sent: 6696; received: 6696; drops: 0
sent: 6893; received: 6891; drops: 2
sent: 7090; received: 7090; drops: 0
sent: 7287; received: 7287; drops: 0
sent: 7484; received: 7484; drops: 0
sent: 7681; received: 7681; drops: 0
***Rate statistics***
Rate: 196.174 DISCOVER-OFFER/second, expected rate: 200

***Statistics for: DISCOVER-OFFER***
sent packets: 7851
received packets: 7847
drops: 4

min delay: 0.430 ms
avg delay: 24.815 ms
max delay: 265.533 ms
std deviation: 42.242 ms
collected packets: 0

```

pallotron commented:
One thing I have noticed is that the perfdhcp -6 doesn't seem to be working:

```
$  sudo ./src/bin/perfdhcp/perfdhcp -6 -r 1 -p 30 -t 1 -i -M ~/test.macs.txt 2401:db00:3010:706b:face:0:1:0
Running: perfdhcp -6 -r 1 -p 30 -t 1 -i -M /home/pallotron/test.macs.txt 2401:db00:3010:706b:face:0:1:0
sent: 0; received: 0; drops: 0
sent: 1; received: 0; drops: 1
^CFailed to receive DHCPv6 packet: Interrupted system call
***Rate statistics***
Rate: 0 Solicit-Advertise/second, expected rate: 1

***Statistics for: SOLICIT-ADVERTISE***
sent packets: 2
received packets: 0
drops: 2

Delay summary unavailable! No packets received.
```

On the server side I only see SOLICIT messages but no response:

```
# tshark udp port 547 and host 2401:db00:3011:26:face:0:1b:0
Running as user "root" and group "root". This could be dangerous.
Capturing on 'eth0'
  1 0.000000000 2401:db00:3011:26:face:0:1b:0 -> 2401:db00:3010:706b:face:0:1:0 DHCPv6 110 Solicit XID: 0x0 CID: 00030001089e01b47209
  2 1.000862397 2401:db00:3011:26:face:0:1b:0 -> 2401:db00:3010:706b:face:0:1:0 DHCPv6 110 Solicit XID: 0x1 CID: 00030001089e01ba2829
  3 2.001891153 2401:db00:3011:26:face:0:1b:0 -> 2401:db00:3010:706b:face:0:1:0 DHCPv6 110 Solicit XID: 0x2 CID: 00030001f45214c25234
```

Shouldn't the packets be RELAY-FWD packets rather than straight SOLICITs?

pallotron commented:
I have noticed that in `TestControl::sendDiscover4` you set relay address:

```
// Set client's and server's ports as well as server's address,
// and local (relay) address.
setDefaults4(socket, pkt4);





void
TestControl::setDefaults4(const TestControlSocket& socket,
                          const Pkt4Ptr& pkt) {
    CommandOptions& options = CommandOptions::instance();
    // Interface name.
    IfacePtr iface = IfaceMgr::instance().getIface(socket.ifindex_);
    if (iface == NULL) {
        isc_throw(BadValue, "unable to find interface with given index");
    }
    pkt->setIface(iface->getName());
    // Interface index.
    pkt->setIndex(socket.ifindex_);
    // Local client's port (68)
    pkt->setLocalPort(DHCP4_CLIENT_PORT);
    // Server's port (67)
    pkt->setRemotePort(DHCP4_SERVER_PORT);
    // The remote server's name or IP.
    pkt->setRemoteAddr(IOAddress(options.getServerName()));
    // Set local address.
    pkt->setLocalAddr(IOAddress(socket.addr_));
    // Set relay (GIADDR) address to local address.
    pkt->setGiaddr(IOAddress(socket.addr_));
    // Pretend that we have one relay (which is us).
    pkt->setHops(1);
}
```

I believe `TestControl::setDefaults6` should be modified to encapsulate SOLICIT pkt6 into a relay-forward... do you agree?

pallotron commented:
Ok I added relayinfo options and also changed the port to 547 and now it works :)

```
$  sudo ./src/bin/perfdhcp/perfdhcp -6 -r 1 -p 30 -t 1 -i -M ~/test.macs.txt 2401:db00:3010:706b:face:0:1:0 -l eth0
Running: perfdhcp -6 -r 1 -p 30 -t 1 -i -M /home/pallotron/test.macs.txt -l eth0 2401:db00:3010:706b:face:0:1:0
sent: 0; received: 0; drops: 0
sent: 1; received: 1; drops: 0
sent: 2; received: 2; drops: 0
^CFailed to receive DHCPv6 packet: Interrupted system call
***Rate statistics***
Rate: 0.857731 Solicit-Advertise/second, expected rate: 1

***Statistics for: SOLICIT-ADVERTISE***
sent packets: 3
received packets: 3
drops: 0

min delay: 13.833 ms
avg delay: 18.431 ms
max delay: 26.949 ms
std deviation: 6.030 ms
collected packets: 0
```

fxdupont commented:
Angelo Failla writes:
> # tshark udp port 547 and host 2401:db00:3011:26:face:0:1b:0
> Running as user "root" and group "root". This could be dangerous.
> Capturing on 'eth0'
>   1 0.000000000 2401:db00:3011:26:face:0:1b:0 -> 2401:db00:3010:706b:face:0:1
> :0 DHCPv6 110 Solicit XID: 0x0 CID: 00030001089e01b47209
>   2 1.000862397 2401:db00:3011:26:face:0:1b:0 -> 2401:db00:3010:706b:face:0:1
> :0 DHCPv6 110 Solicit XID: 0x1 CID: 00030001089e01ba2829
>   3 2.001891153 2401:db00:3011:26:face:0:1b:0 -> 2401:db00:3010:706b:face:0:1
> :0 DHCPv6 110 Solicit XID: 0x2 CID: 00030001f45214c25234
> ```

> Shouldn't the packets be RELAY-FWD packets rather than straight SOLICITs?

=> no but for sure SOLICITs must be multicasted.

Regards

Francis Dupont <fdupont@isc.org>

pallotron commented:
yeah but for a tool like perfdhcp that doesn't work well, anyway, I made some changes, this PR is ready for review. I made v6 tests consistent to the v4 ones. v4 packets are simulating stuff coming from a relayer so I made v6 the same.

fxdupont commented:
Angelo Failla writes:
> yeah but for a tool like perfdhcp that doesn't work well

=> it works well but is very hairy to use and unfortunately the documentation
is not yet enough for newcomers...

> anyway, I made some
>  changes, this PR is ready for review. I made v6 tests consistent to the v4 o
> nes. v4 packets are simulating stuff coming from a relayer so I made v6 the s
> ame.

=> it is not a consistency problem: for v4 there is no choice in particular
for a tool using the standard IP stack, so v4 packets are from a relay,
v6 packets are from what you like.

Regards

Francis Dupont <fdupont@isc.org>

pallotron commented:
Ok, I guess what I need to know is what you think about this patch.
I need to simulate something similar to production traffic so this allows to give the tool a list of real macs (like 1k ones) and have it choose a random one from the list for every request sent.

pallotron commented:
Any chance you can look into this and/or merge this?

vrisk commented:
Hi Angelo - thanks for sharing this patch. I am sorry that nobody has had time to look into this, the whole team is focused on completing the Kea 1.1 server release.  (We have to focus on the stuff that will enable additional deployments as our first priority, to build support for Kea.)  It sounds like you have implemented what you need for your testing, so there is no urgency to us looking at this, is that right?

pallotron commented:
yeah no urgency right now, release is more important :)
i have committed this into our internal repo for now so we can use my modifications.

pallotron commented:
ping? :)

msiodelski commented:
Today I went over the proposed patch and I don't see any major issues with it. I don't see a reason to make a v6 traffic relayed. At some point it may be able to add configurable number of relay encapsulations to test server performance when dealing with decapsulating the relayed messages. But, that may need more extensive changes. I propose that for this pull request we simply adopt the part with MAC addresses. I will make a careful review of this code soon. Again, from the first glance it looks pretty good. Perhaps we'll need to add more extensive tests to make sure that the Mac addresses from the files are indeed used by the perfdhcp.

pallotron commented:
thanks for taking time to review this.

> Today I went over the proposed patch and I don't see any major issues with it. I don't see a reason to make a v6 traffic relayed. At some point it may be able to add configurable number of relay encapsulations to test server performance when dealing with decapsulating the relayed messages. But, that may need more extensive changes.

I explained in my comments above why I need to use relay info.
I am used relay traffic in v4 because this is the same thing you do in v4, and I actually tried this in our infra and it was not working, as soon as I added the encapsulation it started working.

> I propose that for this pull request we simply adopt the part with MAC addresses. 

I think it can go with the relayinfo part.

> I will make a careful review of this code soon. Again, from the first glance it looks pretty good. Perhaps we'll need to add more extensive tests to make sure that the Mac addresses from the files are indeed used by the perfdhcp.

Isn't `CommandOptionsTest.LoadMacsFromFile` test enough? I added it to `src/bin/perfdhcp/tests/command_options_unittest.cc` (see https://github.com/isc-projects/kea/pull/22/files#diff-9f889dcb7f2acd9c1d6c244ba82042cbR839)

msiodelski commented:
On 19.07.2016 00:29, Angelo Failla wrote:
> thanks for taking time to review this.
>
>     Today I went over the proposed patch and I don't see any major
>     issues with it. I don't see a reason to make a v6 traffic relayed.
>     At some point it may be able to add configurable number of relay
>     encapsulations to test server performance when dealing with
>     decapsulating the relayed messages. But, that may need more
>     extensive changes.
>
> I explained in my comments above why I need to use relay info.
> I am used relay traffic in v4 because this is the same thing you do in
> v4, and I actually tried this in our infra and it was not working, as
> soon as I added the encapsulation it started working.
>
>     I propose that for this pull request we simply adopt the part with
>     MAC addresses.
>
> I think it can go with the relayinfo part.
>

The problem you hit seems to be that the server filters out certain
messages sent to unicast (Solicit, Rebind etc), which (by spec) must be
sent to multicast All_DHCP_Relay_Agents_and_Servers address. Did you try
to run your test using multicast? That would mean you'd need to remove
the last parameter from the command line you're using.

>     I will make a careful review of this code soon. Again, from the
>     first glance it looks pretty good. Perhaps we'll need to add more
>     extensive tests to make sure that the Mac addresses from the files
>     are indeed used by the perfdhcp.
>
> Isn't |CommandOptionsTest.LoadMacsFromFile| test enough? I added it to
> |src/bin/perfdhcp/tests/command_options_unittest.cc| (see
> https://github.com/isc-projects/kea/pull/22/files#diff-9f889dcb7f2acd9c1d6c244ba82042cbR839)
>

That merely checks that the MAC addresses are loaded from file, not that
they are used by perfdhcp.


Marcin

msiodelski commented:
# Review

## src/bin/perfdhcp/command_options.cc
The help text:
"-M<mac-file-list>: A text file containing a list of macs, one per line. If provided a random mac will be choosen for every exchange."

should mention that this parameter must not be used in conjuction with -b<base> parameter.

Also, the name of the parameter is inconsistent with its name a couple lines up, which is -M<mac_list_file>. In both cases it should be -M<mac-list-file> (also note hyphens, rather than underscores).

The usage text should also be clear that MAC addresses are used in v6 case to generate DUID-LLs, rather than MAC addresses carried in layer 2.


## src/bin/perfdhcp/command_options.h
getMacListFile description: The brief description should rather say "Returns location of the file containing list of MAC addresses".

And then it should follow with a text saying that MAC addresses read from the file are used by the perfdhcp in message exchanges with DHCP server.

The "template" means something different for perfdhcp. The template is a message template in which some parameters can be randomized.

getAllMacs should be made const:
const std::vector<std::vector<uint8_t> >& getAllMacs() const { return mac_list_; }

Description of mac_list_file_ and mac_list_:
Please use three backslashes for the description of those members to make them doxygen style. Also there is a TODO left for the mac_list_ member.

loadMacs: brief description should start with capital letter.

validateMac: brief description should start with capital letter.

validateMac: line should be passed by const reference, rather than reference.

validateMac should rather be renamed to decodeMac, as validation of merely a side effect of decoding.

Also, we shouldn't be inventing new functions to decode MAC address. I suggest you use HWAddr objec to decode MAC address instead.

Value returned by validateMac should be in parens => return (false);


## src/bin/perfdhcp/test_control.cc

generateMacAddress: return macs[r] => return(mac[r]);

_generateDuid:_ The description of the generateDuid in test_control.h should be extended to say that when the MAC file list is read from a file the generate DUID is DUID-LL. Otherwise the DUID is DUID-LLT.

Please replace this:
```c++
// put the mac address bytes at the end
duid[4] = mac[0];
duid[5] = mac[1];
duid[6] = mac[2];
duid[7] = mac[3];
duid[8] = mac[4];
duid[9] = mac[5];
```

with this:

```c++
#include <algorithm>;
...
std::copy(mac.begin(), mac.end(), duid.begin() + 4);
```

## src/bin/perfdhcp/test_control.h
**number_generator_** has no description.


## src/bin/perfdhcp/tests/command_options_unittest.cc
There is no test that checks that the loaded MAC addresses are actually used. If nothing else the test should at least call TestControl::generateMacAddress and TestControl::generateDuid after loading the file and check that the generated Mac addresses include those that you specified in the file. If there is a risk that in 4 iterations some of them will be generated twice, and other not generated at all (depending how uniform generator behaves) we should at least test that one generated MAC belongs to the group of those in the file.


## src/bin/perfdhcp/tests/testdata/mac_list.txt
Makefile.am in this directory should be updated to include mac_list.txt as EXTRA_DIST. Otherwise, the distcheck is going to fail.

pallotron commented:
Thanks for the detailed review! I will work on it and come back here with changes!


> The problem you hit seems to be that the server filters out certain
messages sent to unicast (Solicit, Rebind etc), which (by spec) must be
sent to multicast All_DHCP_Relay_Agents_and_Servers address. Did you try
to run your test using multicast? That would mean you'd need to remove
the last parameter from the command line you're using.

I can't use multicast in Facebook's production network. I need to specifically point to a given IP. E.g. my devserver (where I run perfchp) is in a different cluster than the target KEA host.
I need to specifically use unicast, hence the relayinfo.

pallotron commented:
ok I added unit tests and implemented the changes requested. this is now ready for review :)

pallotron commented:
ping

msiodelski commented:
I am on holidays until August 16th, so I'll get back to you next week.
Marcin

On 04.08.2016 01:23, Angelo Failla wrote:
> ping
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/22#issuecomment-237405641>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABro_21Vva7dnOjUupb-UjW_X_iWHDCOks5qcSMAgaJpZM4ImdC0>.
>

pallotron commented:
@msiodelski you back from holidays? :) do the last commits w/ unit tests and implemented suggestions satisfy you guys?

msiodelski commented:
I am back from holidays, though, we are currently working hard on
completing Kea 1.1 beta release and I didn't have chance to re-review
your updates.

I should be able to get to it shortly.

Marcin

On 23.08.2016 18:24, Angelo Failla wrote:
> @msiodelski <https://github.com/msiodelski> you back from holidays? :)
> do the last commits w/ unit tests and implemented suggestions satisfy
> you guys?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/22#issuecomment-241790683>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABro_-qP1Ib4Lu3GSh9ucjsOoy381LDYks5qix64gaJpZM4ImdC0>.
>

msiodelski commented:
I reviewed 3 most recent commits on 'perfdhcp_random_mac_from_list' branch and your changes look fine. There are some very little things which I can handle myself when merging.

One thing I'd like you to fix, though, is the issue in new tests. That is the following line:
```c++
std::find(macs.begin(), macs.end(), mac);
```

which is meant to test that the particular MAC address has been read from the file list. In fact, this code merely invokes std::find, but doesn't check the returned value. Note, that on success the returned value should point to an element on the list, rather than to macs.end(). 

There is also a general problem with using relayed traffic, which you desire in your test case. The current patch enforces the use of relayed message. Previously, the message was not relayed. There are people out there currently using perfdhcp in their environments, and what you're proposing is an incompatible change, which can affect their test scenarios. Relayed messages in DHCPv6 are encapsulated (different than in DHCPv4), which means that the server needs to perform additional steps to decapsulate relayed message. That may affect server's performance presented by perfdhcp. Also, I have no guarantee that people using perfdhcp don't have their test environment setup in such a way that relayed traffic will not go through.... similar, but opposite to your problem, whereby your environment doesn't allow for not relayed traffic.

I am not opposed to adding ability of perfdhcp to generate relayed traffic, as this is giving us a chance to verify the performance of a server operating on relayed traffic. But... we have to make it optional feature (controlled by the use of additional parameter). We also should be able to control the level of such encapsulation - up to 32 relays. That way, you can use the new parameter to test server in your environment, but nobody else is affected by this change. And... it comes with additional feature which can be desired by other users.

If you're willing to address that, you're welcome to do so. If not, I can apply this change, but I am iffy about being able to put this into Kea 1.1-beta release as we're constrained by the time. We can also put your patch for MAC address list (but without relayed traffic) into Kea 1.1 beta and add relayed traffic later (when there is more time).

pallotron commented:
Hey msiodelski,
I will address your feedback on the std::find issue.
Also, I totally understand your opinion on making relayed behavior
optional, I will put a commit to make it optional and controlled by an argv
switch :)
On Wed 24 Aug 2016 at 17:25, Marcin Siodelski <notifications@github.com>
wrote:

> I reviewed 3 most recent commits on 'perfdhcp_random_mac_from_list' branch
> and your changes look fine. There are some very little things which I can
> handle myself when merging.
>
> One thing I'd like you to fix, though, is the issue in new tests. That is
> the following line:
>
> std::find(macs.begin(), macs.end(), mac);
>
> which is meant to test that the particular MAC address has been read from
> the file list. In fact, this code merely invokes std::find, but doesn't
> check the returned value. Note, that on success the returned value should
> point to an element on the list, rather than to macs.end().
>
> There is also a general problem with using relayed traffic, which you
> desire in your test case. The current patch enforces the use of relayed
> message. Previously, the message was not relayed. There are people out
> there currently using perfdhcp in their environments, and what you're
> proposing is an incompatible change, which can affect their test scenarios.
> Relayed messages in DHCPv6 are encapsulated (different than in DHCPv4),
> which means that the server needs to perform additional steps to
> decapsulate relayed message. That may affect server's performance presented
> by perfdhcp. Also, I have no guarantee that people using perfdhcp don't
> have their test environment setup in such a way that relayed traffic will
> not go through.... similar, but opposite to your problem, whereby your
> environment doesn't allow for not relayed traffic.
>
> I am not opposed to adding ability of perfdhcp to generate relayed
> traffic, as this is giving us a chance to verify the performance of a
> server operating on relayed traffic. But... we have to make it optional
> feature (controlled by the use of additional parameter). We also should be
> able to control the level of such encapsulation - up to 32 relays. That
> way, you can use the new parameter to test server in your environment, but
> nobody else is affected by this change. And... it comes with additional
> feature which can be desired by other users.
>
> If you're willing to address that, you're welcome to do so. If not, I can
> apply this change, but I am iffy about being able to put this into Kea
> 1.1-beta release as we're constrained by the time. We can also put your
> patch for MAC address list (but without relayed traffic) into Kea 1.1 beta
> and add relayed traffic later (when there is more time).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/22#issuecomment-242126454>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfDH6bFl262dsfGT_pFhq-k7rpvpZQ9ks5qjHCTgaJpZM4ImdC0>
> .
>
-- 

Angelo "pallotron" Failla - pallotron(at)[freaknet.org|gmail.com]

W3: http://www.angelofailla.com

Blog: http://blog.angelofailla.com

FB: https://www.facebook.com/pallotron

TW: https://twitter.com/pallotron

GPG Key ID: 89450920 available @ pgp.mit.edu

GPG FingerPrint: 0DE8 DB22 538E D6B9 8784  83E5 1BEA 7D5C 8945 0920

msiodelski commented:
ACK! Thanks!

On 24.08.2016 20:23, Angelo Failla wrote:
> Hey msiodelski,
> I will address your feedback on the std::find issue.
> Also, I totally understand your opinion on making relayed behavior
> optional, I will put a commit to make it optional and controlled by an argv
> switch :)
> On Wed 24 Aug 2016 at 17:25, Marcin Siodelski <notifications@github.com>
> wrote:
> 
>> I reviewed 3 most recent commits on 'perfdhcp_random_mac_from_list' branch
>> and your changes look fine. There are some very little things which I can
>> handle myself when merging.
>>
>> One thing I'd like you to fix, though, is the issue in new tests. That is
>> the following line:
>>
>> std::find(macs.begin(), macs.end(), mac);
>>
>> which is meant to test that the particular MAC address has been read from
>> the file list. In fact, this code merely invokes std::find, but doesn't
>> check the returned value. Note, that on success the returned value should
>> point to an element on the list, rather than to macs.end().
>>
>> There is also a general problem with using relayed traffic, which you
>> desire in your test case. The current patch enforces the use of relayed
>> message. Previously, the message was not relayed. There are people out
>> there currently using perfdhcp in their environments, and what you're
>> proposing is an incompatible change, which can affect their test
> scenarios.
>> Relayed messages in DHCPv6 are encapsulated (different than in DHCPv4),
>> which means that the server needs to perform additional steps to
>> decapsulate relayed message. That may affect server's performance
> presented
>> by perfdhcp. Also, I have no guarantee that people using perfdhcp don't
>> have their test environment setup in such a way that relayed traffic will
>> not go through.... similar, but opposite to your problem, whereby your
>> environment doesn't allow for not relayed traffic.
>>
>> I am not opposed to adding ability of perfdhcp to generate relayed
>> traffic, as this is giving us a chance to verify the performance of a
>> server operating on relayed traffic. But... we have to make it optional
>> feature (controlled by the use of additional parameter). We also should be
>> able to control the level of such encapsulation - up to 32 relays. That
>> way, you can use the new parameter to test server in your environment, but
>> nobody else is affected by this change. And... it comes with additional
>> feature which can be desired by other users.
>>
>> If you're willing to address that, you're welcome to do so. If not, I can
>> apply this change, but I am iffy about being able to put this into Kea
>> 1.1-beta release as we're constrained by the time. We can also put your
>> patch for MAC address list (but without relayed traffic) into Kea 1.1 beta
>> and add relayed traffic later (when there is more time).
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/isc-projects/kea/pull/22#issuecomment-242126454>,
> or mute
>> the thread
>>
> <https://github.com/notifications/unsubscribe-auth/AAfDH6bFl262dsfGT_pFhq-k7rpvpZQ9ks5qjHCTgaJpZM4ImdC0>
>> .
>>
> -- 
> 
> Angelo "pallotron" Failla - pallotron(at)[freaknet.org|gmail.com]
> 
> W3: http://www.angelofailla.com
> 
> Blog: http://blog.angelofailla.com
> 
> FB: https://www.facebook.com/pallotron
> 
> TW: https://twitter.com/pallotron
> 
> GPG Key ID: 89450920 available @ pgp.mit.edu
> 
> GPG FingerPrint: 0DE8 DB22 538E D6B9 8784 83E5 1BEA 7D5C 8945 0920
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/22#issuecomment-242162272>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABro_47C_TXHyLGMl0wA08nkn0I6TYI_ks5qjIwvgaJpZM4ImdC0>.
>

pallotron commented:
ok all done, checking result of std::find, make dhcpv6 relay behavior non default, added a few unit tests.
please review next week :)

msiodelski commented:
## src/bin/perfdhcp/command_options.cc
The -A parameter should come with an integer parameter that specifies the level of encapsulation of the relayed message, i.e. how many relays forwarded this message. I am fine with supporting only one level of encapsulation at present, but the parameter should require a value from day one. You can put the limitation that the only valid value is 1, and we will extend that in the future to allow values from 1 to 32.

In the help string, you should put the description of the -A parameter into the section of "DHCPv6 only options" as it is only supported for DHCPv6.



## src/bin/perfdhcp/test_control.cc
Avoid using "relayer" in favor of "relay agent".


## src/bin/perfdhcp/tests/test_control_unittest.cc
Instead of using:
```c++
ASSERT_NE(std::find(macs.begin(), macs.end(), mac), macs.end());
```

use

```c++
ASSERT_FALSE(std::equal(...));
```

because ASSERT_NE requires that there are << operator overloads for the parameters being compared. For a vector it will cause build issues on some sytems.


There are no tests that verify that the generated messages are indeed relayed when the -A parameter is specified.

pallotron commented:
Addressed feedbacks in my last commit (e3f6d51)

msiodelski commented:
ok, thanks. I'll review shortly.

On 30.08.2016 17:24, Angelo Failla wrote:
> Addressed feedbacks in my last commit (e3f6d51
> <https://github.com/isc-projects/kea/commit/e3f6d512e16b951d2457bed02b3ff006b3cab1ba>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/22#issuecomment-243477357>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABro_5A6Ngjtg0A4DPgNh9_qSQGEKMwhks5qlEsjgaJpZM4ImdC0>.
>

msiodelski commented:
I have integrated your changes into new branch github22 in our Kea github repo: https://github.com/isc-projects/kea.git.

Can you please checkout this branch and see if the new functionality works ok in your environment? Note that I have applied several (minor) updates, so functionally there should be no change, but better be on the safe side before we merge this to master.

Thanks,
Marcin

pallotron commented:
Sure, will check tomorrow
On Wed 31 Aug 2016 at 19:01, Marcin Siodelski <notifications@github.com>
wrote:

> I have integrated your changes into new branch github22 in our Kea github
> repo: https://github.com/isc-projects/kea.git.
>
> Can you please checkout this branch and see if the new functionality works
> ok in your environment? Note that I have applied several (minor) updates,
> so functionally there should be no change, but better be on the safe side
> before we merge this to master.
>
> Thanks,
> Marcin
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/22#issuecomment-243848570>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfDH3bmskuItMUvLIjFV_e0VuReHKGYks5qlcGDgaJpZM4ImdC0>
> .
>
-- 

Angelo "pallotron" Failla - pallotron(at)[freaknet.org|gmail.com]

W3: http://www.angelofailla.com

Blog: http://blog.angelofailla.com

FB: https://www.facebook.com/pallotron

TW: https://twitter.com/pallotron

GPG Key ID: 89450920 available @ pgp.mit.edu

GPG FingerPrint: 0DE8 DB22 538E D6B9 8784  83E5 1BEA 7D5C 8945 0920

pallotron commented:
Hi @msiodelski 

I checked out github22 branch and tested it in our environment and it works:

```
[pallotron@devXXX.lla1.facebook.com] (branch: github22)
=> ~/github/kea/src/bin/perfdhcp
$ git branch
* github22
  master
  perfdhcp_random_mac_from_list
[Exit code 0 @ 02:04:05]

[pallotron@devXXX.lla1.facebook.com] (branch: github22)
=> ~/github/kea/src/bin/perfdhcp
$ sudo ./perfdhcp -R 1 -6 -r 1 -p 30 -t 1 -i -A1 -M ~/test.macs.txt 2401:db00:3010:e063:face:0:7:0
Running: perfdhcp -R 1 -6 -r 1 -p 30 -t 1 -i -A 1 -M /home/pallotron/test.macs.txt 2401:db00:3010:e063:face:0:7:0
sent: 0; received: 0; drops: 0
sent: 1; received: 1; drops: 0
sent: 2; received: 2; drops: 0
sent: 3; received: 3; drops: 0
^CFailed to receive DHCPv6 packet: Interrupted system call
***Rate statistics***
Rate: 0.637356 Solicit-Advertise/second, expected rate: 1

***Statistics for: SOLICIT-ADVERTISE***
sent packets: 4
received packets: 4
drops: 0

min delay: 10.503 ms
avg delay: 11.685 ms
max delay: 13.594 ms
std deviation: 1.362 ms
collected packets: 0
```

msiodelski commented:
Thanks for verifying the code and obviously for working on the new features. I integrated those changes into master branch. I will be available in the Kea 1.1.0 release later this month.

pallotron commented:
ty!
